### PR TITLE
Only export #[no_mangle] extern symbols during LTO

### LINKED ISSUE #34985: LTO preserves too many `extern` symbols
Right now the standard library ends up exporting a bunch of `extern` symbols from liblibc:

```
$ echo '#[no_mangle] pub extern fn foo() { panic!() }' > foo.rs
$ echo 'extern void foo(); int main() { foo(); return 0; }' > foo.c
$ rustc foo.rs --crate-type staticlib -C lto
$ gcc foo.c libfoo.a
$ nm -g ./a.out | grep libc
000000000049e310 T _ZN4libc4unix6notbsd11WEXITSTATUS17hdf8e9951afe1c101E
000000000049e530 T _ZN4libc4unix6notbsd5linux7CPU_SET17hb5ba54c4f04e4ad5E
000000000049e510 T _ZN4libc4unix6notbsd5linux8CPU_ZERO17hc3389841ca5535cfE
000000000049e630 T _ZN4libc4unix6notbsd5linux9CPU_EQUAL17h6aa5067ba806ec07E
000000000049e5b0 T _ZN4libc4unix6notbsd5linux9CPU_ISSET17h25c9bc3ffd366857E
000000000049e170 T _ZN4libc4unix6notbsd6FD_CLR17h82439e27c061361cE
000000000049e250 T _ZN4libc4unix6notbsd6FD_SET17h6f0001978b943490E
000000000049e2c0 T _ZN4libc4unix6notbsd7FD_ZERO17h157d16767431163aE
000000000049e1e0 T _ZN4libc4unix6notbsd8FD_ISSET17h9bb2d4903026e75fE
000000000049e320 T _ZN4libc4unix6notbsd8WTERMSIG17h140bc8346daf707fE
000000000049e2f0 T _ZN4libc4unix6notbsd9WIFEXITED17h9b79d0f36c3cdb42E
```

These are all defined in the libc crate, but LTO should be stripping them out.

I believe the reason that this is happening is we attempt to preserve `extern` symbols from upstream crates (as they may be part of the ABI), but we don't check for `#[no_mangle]` (which none of these functions have).

I think the relevant clause [is here](https://github.com/rust-lang/rust/blob/ad264f7f39b6113fb87a489d72d31abbc03ab148/src/librustc_trans/base.rs#L2639-L2654) which ends up calling [this function](https://github.com/rust-lang/rust/blob/ad264f7f39b6113fb87a489d72d31abbc03ab148/src/librustc_metadata/csearch.rs#L276) coming over to [this function](https://github.com/rust-lang/rust/blob/ad264f7f39b6113fb87a489d72d31abbc03ab148/src/librustc_metadata/decoder.rs#L1534-L1562). Although there's [a clause](https://github.com/rust-lang/rust/blob/ad264f7f39b6113fb87a489d72d31abbc03ab148/src/librustc_metadata/decoder.rs#L1557-L1558) which [looks like it checks for this](https://github.com/rust-lang/rust/blob/ad264f7f39b6113fb87a489d72d31abbc03ab148/src/libsyntax/attr.rs#L339-L342), I think we're [falling through too quickly](https://github.com/rust-lang/rust/blob/ad264f7f39b6113fb87a489d72d31abbc03ab148/src/librustc_metadata/decoder.rs#L1549) and we may wish to change that.

Note that this is mostly just a hunch though, some more investigation may be required! For an aspiring trans hacker though I don't think this should be *too* hard and I'd be willing to help out!

### PR DESCRIPTION
**Pull request by wesleywiser**
Fixes #34985

### CODE REVIEW

File: src/librustc_metadata/decoder.rs [hunk 1]
```diff
@@ -1542,13 +1541,9 @@ pub fn is_extern_item<'a, 'tcx>(cdata: Cmd,
     let applicable = match item_family(item_doc) {
         ImmStatic | MutStatic => true,
         Fn => {
-            let ty::TypeScheme { generics, ty } = get_type(cdata, id, tcx);
+            let ty::TypeScheme { generics, ty: _ } = get_type(cdata, id, tcx);
```
Comment (alexcrichton) [inline src/librustc_metadata/decoder.rs]: You should be able to actually change this to:

```rust
let ty::TypeScheme { generics, _ } = ...
```

that is, remove the `ty` matching entirely

**Comment by alexcrichton** [comment]
Thanks @wesleywiser! The PR/commit may want to be updated to indicate though that only extern + no_mangle symbols are exported across LTO. I believe today we basically export everything with extern, whether or not it has no_mangle.

I'd love to write a test for this as well, but I have no idea how to do so, so seems fine to just have manual verification :)

**Comment by alexcrichton** [comment]
Also just to confirm, could you gist the contents of the commands in https://github.com/rust-lang/rust/issues/34985 after this PR?

File: src/librustc_metadata/decoder.rs (hunk 1 above)
Comment (wesleywiser) [inline src/librustc_metadata/decoder.rs]: Thank you! I was trying to find that syntax but couldn't.

**Comment by wesleywiser** [comment]
The output of the commands in #34985 is now empty. Did you want the full output of `nm` without the `grep`?

**Comment by alexcrichton** [comment]
Nah that's good enough for me! Mind updating the commit message to something like:

> Only export #[no_mangle] extern symbols during LTO
>
> Previously ... (more words here)

**Comment by wesleywiser** [comment]
How's that?

**Comment by alexcrichton** [comment]
@bors: r+ 2dbf00e2bb60b6b3172002c7e00d516c94d9ee42

Looks good to me, thanks @wesleywiser! (also updated the PR title)

**Comment by wesleywiser** [comment]
Thank you!

**Comment by alexcrichton** [comment]
@bors: retry

* https://github.com/rust-lang/rust/issues/33391

**Pull request merged**